src/lib.rs

```diff
@@ -209,3 +211,114 @@ pub extern "C" fn dalek_curved25519_scalarmult_basepoint(pk: *mut u8, e: *mut u8
     // Write back to pk
     pk.copy_from_slice(u.as_bytes());
 }
+
+
+/// Scalar multiplication using the provided basepoint
+#[no_mangle]
+pub extern "C" fn dalek_curve25519_scalarmult(o: *mut PublicKey, e: *mut SecretKey, bp: *mut PublicKey) {
+    let (o, e, bp) = unsafe {( &mut (*o), &(*e), &(*bp) )};
+
+    // Copy secret into editable slice
+    let mut ec = [0u8; 32];
+    ec.copy_from_slice(e);
+
+    let mut bpc = [0u8; 32];
+    bpc.copy_from_slice(bp);
+
+    // Clamp secret key and expand
+    ec[0] &= 248;
+    ec[31] &= 127;
+    ec[31] |= 64;
+
+    let secret = curve25519_dalek::scalar::Scalar::from_bytes_mod_order(ec.clone());
+
+    let basepoint = curve25519_dalek::scalar::Scalar::from_bytes_mod_order(bpc.clone());
+
+    // Perform multiplication
```

Comment:
@isislovecruft iirc we talked about `curve25519_scalarmult` being an `x25519` equivalent but i haven't got my head around what it is trying to do, would appreciate any guidance / corrections here?